src/runtime/prepareSwapTransactions.ts

```diff
@@ -0,0 +1,200 @@
+import { z } from 'zod'
+import { tokenAmountWithMetadata, Transaction } from '../types/shortcuts'
+import { EvmContractCall, Hook as SquidHook } from '@0xsquid/squid-types'
+import { NetworkId } from '../types/networkId'
+import { Address, encodeFunctionData, erc20Abi, parseUnits } from 'viem'
+import {
+  simulateTransactions,
+  UnsupportedSimulateRequest,
+} from './simulateTransactions'
+import { logger } from '../log'
+import { getConfig } from '../config'
+import got, { HTTPError } from 'got'
+import { getClient } from './client'
+
+interface SwapTransaction {
+  swapType: 'same-chain' // only supporting same-chain swaps for now
+  chainId: number
+  buyAmount: string
+  sellAmount: string
+  buyTokenAddress: string
+  sellTokenAddress: string
+  // be careful -- price means different things when using sellAmount vs buyAmount
+  price: string
+  guaranteedPrice: string
+  appFeePercentageIncludedInPrice: string | undefined
+  /**
+   * In percentage, between 0 and 100
+   */
+  estimatedPriceImpact: string | null
+  gas: string
+  estimatedGasUse: string | null | undefined
+  to: Address
+  value: string
+  data: `0x${string}`
+  from: Address
+  allowanceTarget: Address
+}
+
+type GetSwapQuoteResponse = {
+  unvalidatedSwapTransaction?: SwapTransaction
+  details: {
+    swapProvider?: string
+  }
+  errors: {
+    provider: string
+    error: {
+      message: string
+      details: unknown
+    }
+  }[]
+}
+
+export async function prepareSwapTransactions({
+  swapFromToken,
+  postHook,
+  swapToAddress,
+  networkId,
+  walletAddress,
+  simulatedGasPadding,
+}: {
+  swapFromToken: z.infer<typeof tokenAmountWithMetadata>
+  postHook: Omit<
+    SquidHook,
+    'fundAmount' | 'fundToken' | 'provider' | 'logoURI' | 'calls'
+  > & { calls: EvmContractCall[] } // we don't support CosmosCall
+  swapToAddress: Address
+  networkId: NetworkId
+  walletAddress: Address
+  simulatedGasPadding?: bigint[]
+}): Promise<Transaction[]> {
+  let postHookWithSimulatedGas = postHook
+
+  try {
+    const simulatedTransactions = await simulateTransactions({
+      networkId,
+      transactions: postHook.calls.map((call) => ({
+        networkId,
+        from: walletAddress,
+        to: call.target,
+        data: call.callData,
+      })),
+    })
+
+    postHookWithSimulatedGas = {
+      ...postHook,
+      calls: postHook.calls.map((call, index) => {
+        return {
+          ...call,
+          estimatedGas: (
+            BigInt(simulatedTransactions[index].gasNeeded) +
+            (simulatedGasPadding?.[index] ?? 0n)
+          ).toString(),
+        }
+      }),
+    }
+  } catch (error) {
+    if (!(error instanceof UnsupportedSimulateRequest)) {
+      logger.warn(error, 'Unexpected error during simulateTransactions')
+    }
+    // use default already set in the postHook, no changes needed
```

Comment:
nit: seems better to set the default here or comment on https://github.com/valora-inc/hooks/pull/539/files#diff-57288d8f4298dae89c637e7276d8497e62e59a1035d846326f93a8ff855d19cdR84 that it needs to the last statement in the try block.  otherwise, like, someone might add additional statements to the try block that update postHookWithSimulatedGas (so postHookWithSimulatedGas gets set, but then needs updates, but then we throw and don't use the default).